### PR TITLE
feat!: compile artifacts into target-specific directories

### LINKED ISSUE #815: Compiler output structure
If I run:

```sh
wing compile -t sim file1.w
```

but also:

```sh
wing compile -t sim file2.w
```

I'll end up with the following files:

```
cwd/
  target/
    .wing/
    file1.w
    file2.w
    tree.json # this file has been overwritten by the second compilation
  file1.w
  file2.w
```

Is it a know issue? Do we want to handle this?

Related to https://github.com/winglang/wing/pull/770.

### PR DESCRIPTION
**Pull request by Chriscbr**
This PR updates the CLI so that all compiler artifacts (like terraform files, simulator files, etc.) are now synthesized into target-specific directories (closes #815). For example, `wing compile -t tf-aws hello.w` will generate terraform artifacts inside `target/hello.tfaws/` while `wing compile -t tf-azure` will generate terraform artifacts inside `target/hello.tfazure/`. This avoids some problems users had where compiling to different targets would override previously compiled artifacts.

The simulator target now generates a single `.wsim` which is generated directly inside `target/`. (Under the hood, the `.wsim` file now contains the `tree.json` file that was previously generated outside).

This is a breaking change as it may require using a new version of the Wing console that understands how to read the updated .wsim file.

*By submitting this pull request, I confirm that my contribution is made under the terms of the 
[Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

**Comment by Chriscbr** [comment]
FYI @ainvoner @skyrpex @polamoros  I'm labeling this with `do-not-merge` until the console is ready to support these new changes. The only change that affects the console AFAIK is that the `.wsim` file now contains `tree.json` inside it directly (it is no longer generated outside).

File: libs/wingsdk/src/core/app.ts [hunk 1]
```diff
@@ -99,17 +99,46 @@ export class CdktfApp extends Construct implements IApp {
...
+  /**
+   * Move files from `outdir/cdktf.out/stacks/root` to `outdir`.
+   */
+  private moveCdktfArtifactsToOutdir(): void {
+    const cdktfOutdir = this.cdktfApp.outdir;
+    const cdktfStackDir = join(cdktfOutdir, "stacks", TERRAFORM_STACK_NAME);
+
+    const files = readdirSync(cdktfStackDir, { withFileTypes: true });
+    for (const file of files) {
+      if (file.isFile()) {
```
Comment (eladb) [inline libs/wingsdk/src/core/app.ts:135]: TIL

File: libs/wingsdk/src/target-sim/app.ts [hunk 2]
```diff
@@ -45,6 +45,9 @@ export class App extends Construct implements core.IApp {
     // write simulator.json file into workdir
     this.synthSimulatorFile(workdir);
 
+    // write tree.json file into workdir
+    core.synthesizeTree(this, workdir);
```
Comment (eladb) [inline libs/wingsdk/src/target-sim/app.ts:49]: Is this going to break the Wing Console?

Comment (Chriscbr) [reply libs/wingsdk/src/target-sim/app.ts:49]: Probably. But I think it should be possible to update the console to work in a backwards-compatible way

Comment (eladb) [reply libs/wingsdk/src/target-sim/app.ts:49]: Backward compatibility is the responsibility of the "server". In our case it's the wing toolchain, not the Console.

The "right thing" to do here is to emit both the old format and the new format until the console is updated to support the new format.

Comment (Chriscbr) [reply libs/wingsdk/src/target-sim/app.ts:49]: Ok, good point. 🙂 Updated to be backwards compatible... I tested out the changes with the latest version of the Wing console and it works as expected.

**Comment by Chriscbr** [comment]
@eladb Even if we're not breaking the wing console I'm guessing we should still make this a minor version bump, yeah? We're still changing where users are going to find their terraform files etc.

**Review by eladb — APPROVED** [review]

**Pull request merged**